### PR TITLE
Fix implementation of Dijkstra's algorithm

### PR DESCRIPTION
**Pull request by thomash-acinq**
Ensure that we only visit each node once in Dijkstra's algorithm.

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala [hunk 1]
```diff
@@ -208,8 +209,9 @@ object Graph {
     while (toExplore.nonEmpty && !targetFound) {
       // node with the smallest distance from the target
       val current = toExplore.dequeue() // O(log(n))
-      if (current.key != sourceNode) {
-        val currentWeight = bestWeights(current.key) // NB: there is always an entry for the current in the 'bestWeights' map
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala:212]: Are you sure about that change?
At first glance getting the best weight from this map seems useful, doesn't it?

**Review by t-bast — COMMENTED** [review]
Good catch!

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala (hunk 1 above)
Comment (thomash-acinq) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala:212]: It's exactly the same weight. We already have it in `current.weight`, there is no point in getting it from somewhere else.

Actually we could remove `bestWeights` completely, it's not needed by the algorithm, the comparison between the different weights for a node is done inside the priority queue.
Now it's just used to know when to update `bestEdges` and it sometimes prevents adding useless weights to the priority queue.

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala:212]: > the comparison between the different weights for a node is done inside the priority queue.

Right, I glossed over that point, once you realize this it's obviously not needed.

**Review by t-bast — APPROVED** [review]

**Pull request merged**